demos/defender/aws_iot_demo_defender.c

```diff
@@ -18,124 +17,63 @@
-#include "platform/iot_network_freertos.h"
 
 /* Defender includes. */
 #include "aws_iot_defender.h"
 
 /* Includes for initialization. */
 #include "iot_mqtt.h"
-
-/* Set to 1 to enable this demo to connect to echo server.
- * Then in the demo output, it is expected to see one more established TCP connection.
- */
-#define _DEMO_WITH_SOCKET_CONNECTED_TO_ECHO_SERVER    0
-
-static IotNetworkServerInfo_t _DEFENDER_SERVER_INFO = AWS_IOT_NETWORK_SERVER_INFO_AFR_INITIALIZER;
-static IotNetworkCredentials_t _AWS_IOT_CREDENTIALS = AWS_IOT_NETWORK_CREDENTIALS_AFR_INITIALIZER;
-
-#if _DEMO_WITH_SOCKET_CONNECTED_TO_ECHO_SERVER == 1
-    static Socket_t _createSocketToEchoServer();
-#endif
+#include "platform/iot_metrics.h"
 
 /**
- * @brief Runs the defender demo.
+ * @brief The keep-alive interval used for this demo.
  *
- * @return AWS_IOT_DEFENDER_SUCCESS on success, otherwise an error code indicating
- *         the cause of error.
+ * An MQTT ping request will be sent periodically at this interval.
  */
-static AwsIotDefenderError_t _defenderDemo( void );
+#define KEEP_ALIVE_SECONDS    ( 60 )
 
 /**
- * @brief Starts the defender agent.
- *
- * @return AWS_IOT_DEFENDER_SUCCESS on success, otherwise an error code indicating
- *         the cause of error.
+ * @brief The timeout for Defender and MQTT operations in this demo.
  */
-static AwsIotDefenderError_t _startDefender( void );
+#define TIMEOUT_MS            ( 5000 )
+
+/*-----------------------------------------------------------*/
 
 /**
  * @brief Callback used to get notification of defender's events.
+ *
+ * @param[in] pCallbackContext context pointer passed by the application
+ * when callback is regiested in AwsIotDefender_Start()
+ *
+ * @param[im] pointer to AwsIotDefenderCallbackInfo_t containing status of
+ * publish
  */
-static void _defenderCallback( void * param1,
-                               AwsIotDefenderCallbackInfo_t * const pCallbackInfo );
-
-/*-----------------------------------------------------------*/
-
-int vStartDefenderDemo( bool awsIotMqttMode,
-                        const char * pIdentifier,
-                        void * pNetworkServerInfo,
-                        void * pNetworkCredentialInfo,
-                        const IotNetworkInterface_t * pNetworkInterface )
+static void _defenderCallback( void * pCallbackContext,
+                               AwsIotDefenderCallbackInfo_t * const pCallbackInfo )
 {
-    int status = EXIT_FAILURE;
-    IotMqttError_t mqttInitStatus;
-    AwsIotDefenderError_t defenderResult;
-
-    /* Unused parameters */
-    ( void ) awsIotMqttMode;
-    ( void ) pIdentifier;
-    ( void ) pNetworkServerInfo;
-    ( void ) pNetworkCredentialInfo;
-    ( void ) pNetworkInterface;
+    ( void ) pCallbackContext;
 
-    /* Initialize the MQTT library. */
-    mqttInitStatus = IotMqtt_Init();
+    IotLogInfo( "User's callback is invoked on event: %s.", AwsIotDefender_EventType( pCallbackInfo->eventType ) );
```

Comment:
A null check would be required before pCallbackInfo is used

Reply:
pCallbackInfo is populated by the library :
https://github.com/aws/amazon-freertos/pull/1345/files#diff-07a09f90dafd00ac812be087a5f9e378R569
So it will not be NULL.
But it is good practice.